### PR TITLE
Fixes #19 #20 #21 #44 #45

### LINKED ISSUE #19: Manage a Cypher template
https://meet.google.com/linkredirect?authuser=0&dest=https%3A%2F%2Fmedium.com%2Fneo4j%2F5-tips-tricks-for-fast-batched-updates-of-graph-structures-with-neo4j-and-cypher-73c7f693c8cc

### PR DESCRIPTION
**Pull request by conker84**
Also contains #44 and #45

### CODE REVIEW

File: consumer/src/main/kotlin/streams/StreamsTopicService.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import java.lang.IllegalArgumentException
+
+
+private const val STREAMS_TOPIC_KEY: String = "streams.sink.topic."
+
+class StreamsTopicService(private val db: GraphDatabaseAPI, private val streamsSinkConfiguration: StreamsSinkConfiguration) {
+    private val properties: GraphPropertiesProxy = db.dependencyResolver.resolveDependency(EmbeddedProxySPI::class.java).newGraphPropertiesProxy()
+
+    fun set(topic: String, query: String) {
+        db.beginTx().use {
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/StreamsTopicService.kt]: we have to make sure to only call this on a writeable instance.

File: consumer/src/main/kotlin/streams/StreamsTopicService.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            properties.setProperty("$STREAMS_TOPIC_KEY$topic", query)
+            it.success()
+        }
+    }
+
+    fun get(topic: String): String {
+        return db.beginTx().use {
+            if (!properties.hasProperty("$STREAMS_TOPIC_KEY$topic")) {
+                if (streamsSinkConfiguration.topics.containsKey(topic)) {
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/StreamsTopicService.kt]: I don't think we should do this
we should write them at startup if they are missing or are different
but not write during get
which also will fail on clusters

File: consumer/src/main/kotlin/streams/StreamsTopicService.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                    set(topic, streamsSinkConfiguration.topics[topic]!!)
+                } else {
+                    throw IllegalArgumentException("Cypher query not registered for topic $topic")
+                }
+            }
+            return properties.getProperty("$STREAMS_TOPIC_KEY$topic").toString()
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/StreamsTopicService.kt]: just pass null as default and handle it higher up if it's null

File: consumer/src/main/kotlin/streams/StreamsTopicService.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import org.neo4j.kernel.internal.GraphDatabaseAPI
+import java.lang.IllegalArgumentException
+
+
+private const val STREAMS_TOPIC_KEY: String = "streams.sink.topic."
+
+class StreamsTopicService(private val db: GraphDatabaseAPI, private val streamsSinkConfiguration: StreamsSinkConfiguration) {
+    private val properties: GraphPropertiesProxy = db.dependencyResolver.resolveDependency(EmbeddedProxySPI::class.java).newGraphPropertiesProxy()
+
+    fun set(topic: String, query: String) {
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/StreamsTopicService.kt]: do we expose these 3 operations via procedures?

Comment (jexp) [reply consumer/src/main/kotlin/streams/StreamsTopicService.kt]: perahps create an issue

File: consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    constructor(logService: LogService,
+                config: Config,
+                db: GraphDatabaseAPI,
+                availabilityGuard: AvailabilityGuard): super(logService, config, db, availabilityGuard) {
+        this.log = logService.getUserLog(KafkaEventSink::class.java)
+        log.info("Initializing Kafka Sink Connector")
+        this.kafkaConfig = KafkaSinkConfiguration.from(config)
+        this.kafkaConsumer = KafkaConsumer(kafkaConfig.asProperties())
+        this.availabilityGuard = availabilityGuard
+        this.streamsTopicService = StreamsTopicService(db, kafkaConfig.streamsSinkConfiguration)
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt]: Log it.

File: consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        this.queryExecution = StreamsEventSinkQueryExecution(this.streamsTopicService, db)
+    }
+
+    override fun start() {
+        job = GlobalScope.launch(Dispatchers.IO) {
+            try {
+                availabilityGuard.await(30000)
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt]: instead use the availability interface

File: consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                log.info("Kafka Sink Connector started.")
+                if (log.isDebugEnabled) {
+                    log.debug("Subscribing sink topics ${kafkaConfig.streamsSinkConfiguration.topics}")
+                } else {
+                    log.info("Subscribing sink topics ${streamsTopicService.getTopics()}")
+                }
+                kafkaConsumer.subscribe(streamsTopicService.getTopics())
+                while (true) {
+                    val records = kafkaConsumer.poll(kafkaConfig.streamsSinkConfiguration.sinkPollingInterval)
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt]: is this blocking?

File: consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            try {
+                availabilityGuard.await(30000)
+                log.info("Kafka Sink Connector started.")
+                if (log.isDebugEnabled) {
+                    log.debug("Subscribing sink topics ${kafkaConfig.streamsSinkConfiguration.topics}")
+                } else {
+                    log.info("Subscribing sink topics ${streamsTopicService.getTopics()}")
+                }
+                kafkaConsumer.subscribe(streamsTopicService.getTopics())
+                while (true) {
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt]: can we extract this into 2-3 methods?

File: consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+class KafkaEventSinkIT {
+    companion object {
+        /**
+         * Kafka TestContainers uses Confluent OSS images.
+         * We need to keep in mind which is the right Confluent Platform version for the Kafka version this project uses
+         *
+         * Confluent Platform | Apache Kafka
+         *                    |
+         * 4.0.x	          | 1.0.x
```
Comment (jexp) [inline consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt]: should we have a parameterized test for all 3 platforms?

File: consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        val dataProperties = mapOf("prop1" to "foo", "bar" to 1)
+        val data = mapOf("id" to 1, "properties" to dataProperties)
+        val job = GlobalScope.launch {
+            var partition = ThreadLocalRandom.current().nextInt(1)
+            var producerRecord = ProducerRecord(topics[0], partition, System.currentTimeMillis(), 1L,
+                    objectMapper.writeValueAsBytes(data))
+            producer.send(producerRecord)
+        }
+        job.join()
+        delay(3000)
```
Comment (jexp) [inline consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt]: what are we waiting for?

File: consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        db.beginTx().use {
+            db.execute("MATCH (n:Label) RETURN n")
+                    .columnAs<Node>("n").use {
+                        assertEquals(true, it.hasNext())
```
Comment (jexp) [inline consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt]: we can just run the exact query we're looking for and check the count

MATCH (n:Label) WHERE properties(n) = $props RETURN count(*)

Comment (jexp) [inline consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt]: doesn't check that it's only one record

File: consumer/src/test/kotlin/streams/StreamsEventSinkQueryExecutionTest.kt [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+class StreamsEventSinkQueryExecutionTest {
+    private lateinit var db: GraphDatabaseService
+    private lateinit var streamsEventSinkQueryExecution: StreamsEventSinkQueryExecution
+
+    @Before
+    fun setUp() {
+        db = TestGraphDatabaseFactory()
+                .newImpermanentDatabaseBuilder()
+                .newGraphDatabase()
+        val kafkaConfig = KafkaSinkConfiguration(streamsSinkConfiguration = StreamsSinkConfiguration(topics = mapOf("shouldWriteCypherQuery" to "MERGE (n:Label {id: event.id})\n" +
```
Comment (jexp) [inline consumer/src/test/kotlin/streams/StreamsEventSinkQueryExecutionTest.kt]: do we also have a plain StreamsSinkConfiguration test?

File: consumer/src/test/kotlin/streams/StreamsTopicServiceTest.kt [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    private lateinit var streamsTopicService: StreamsTopicService
+
+    private lateinit var kafkaConfig: KafkaSinkConfiguration
+
+    @Before
+    fun setUp() {
+        db = TestGraphDatabaseFactory()
+                .newImpermanentDatabaseBuilder()
+                .newGraphDatabase()
+        kafkaConfig = KafkaSinkConfiguration(streamsSinkConfiguration = StreamsSinkConfiguration(topics = mapOf("shouldWriteCypherQuery" to "MERGE (n:Label {id: event.id})\n" +
```
Comment (jexp) [inline consumer/src/test/kotlin/streams/StreamsTopicServiceTest.kt]: I wonder if we should have some "special" query tokens like `CREATE_NODE_FROM_TOPIC_EVENT` etc.
But that's perhaps a stupid idea and we should keep it explicit.

File: readme.adoc [hunk 14]
```diff
@@ -10,86 +10,16 @@ The project is composed by 2 parts:
...
 
 ----
 mvn clean install
 ----
 
-2. Copy `target/neo4j-kafka-*.jar` into `$NEO4J_HOME/plugins`
+2. Copy `<producer_or_consumer_dir>/target/neo4j-streams-*.jar` into `$NEO4J_HOME/plugins`
 3. Restart Neo4j
 
-== Quick Configuration
```
Comment (jexp) [inline readme.adoc]: Where does the user see the quick config in the readme?
I wouldn't remove it.

File: consumer/src/main/kotlin/streams/StreamsTopicService.kt (hunk 4 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/StreamsTopicService.kt]: > perahps create an issue

Is the #46, is based on what we have talked about on Slack. Give it a check if you want update it

File: consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt (hunk 7 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt]: Yes, it blocks until new records arrive at the consumer, it's placed inside a Couroutine so all the work is delegated to another thread.

File: consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt (hunk 9 above)
Comment (conker84) [reply consumer/src/test/kotlin/integrations/KafkaEventSinkIT.kt]: created #51

File: consumer/src/test/kotlin/streams/StreamsEventSinkQueryExecutionTest.kt (hunk 12 above)
Comment (conker84) [reply consumer/src/test/kotlin/streams/StreamsEventSinkQueryExecutionTest.kt]: Done

File: readme.adoc (hunk 14 above)
Comment (conker84) [reply readme.adoc]: Added to links in order to don't made the same changes on different files ok?

File: consumer/src/test/kotlin/streams/StreamsTopicServiceTest.kt (hunk 13 above)
Comment (conker84) [reply consumer/src/test/kotlin/streams/StreamsTopicServiceTest.kt]: I think we could leave this possibility to the end user, being the template configurable the user can inject fixed params in the query. Btw I can open an issue if you want

File: consumer/src/main/kotlin/streams/StreamsTopicService.kt (hunk 1 above)
Comment (jexp) [reply consumer/src/main/kotlin/streams/StreamsTopicService.kt]: there is only 1 commit attached to this PR perhaps you pushed to a wrong branch?

**Pull request merged**